cedar-policy-validator/src/lib.rs

```diff
@@ -206,6 +286,192 @@ impl Validator {
         typecheck.typecheck_policy(t, &mut errors, &mut warnings);
         (errors.into_iter(), warnings.into_iter())
     }
+
+    #[cfg(feature = "level-validate")]
+    /// Check that `t` respects `max_allowed_level`
+    /// This assumes that (strict) typechecking has passed
+    fn check_entity_deref_level<'a>(
+        &'a self,
+        t: &'a Template,
+        max_allowed_level: &EntityDerefLevel,
+        policy_id: &PolicyID,
+    ) -> Vec<ValidationError> {
+        let typechecker = Typechecker::new(&self.schema, ValidationMode::Strict, t.id().clone());
+        let type_annotated_asts = typechecker.typecheck_by_request_env(t);
+        let mut errs = vec![];
+        for (_, policy_check) in type_annotated_asts {
+            match policy_check {
+                PolicyCheck::Success(e) => {
+                    let res =
+                        self.check_entity_deref_level_helper(&e, max_allowed_level, policy_id);
+                    match res.1 {
+                        Some(e) => errs.push(ValidationError::EntityDerefLevelViolation(e)),
+                        None => (),
+                    }
+                }
+                PolicyCheck::Fail(e) => println!("Typechecking failed: {:?}", e),
+                PolicyCheck::Irrelevant(_) => (), //Don't report level violations if typechecking already failed
+            }
+        }
+        errs
+    }
+
+    #[cfg(feature = "level-validate")]
+    fn min(
```

Comment:
The minimum of an empty list is infinity?

Reply:
I think we can change this to take a non-empty vec and remove a potential footgun in the future

Reply:
I started implementing that, but I think the current version is actually less error-prone. If we add more support for empty sets / extension functions with no args in the future, this code would handle that correctly, while `NonEmpty::from_vec(v).unwrap()` would panic. I don't want to add actual error handling for these cases that are currently impossible. Marking these expressions as non-dereferenceable should be future-proof (baring some possbile weird extension functions I haven't thought of).